### PR TITLE
Template actions user input validation

### LINKED ISSUE #522: Request input binding and validation
We'll need to add user input validation on all of our action endpoints in a standardized way.

Use gin's model binding and validation as described here: https://github.com/gin-gonic/gin#model-binding-and-validation

### PR DESCRIPTION
**Pull request by Dzalevski**
# Pull Request Template

## Description

Validating user input on 
PutTemplate
PostTemplate 

Fixes #522

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A - Tested with insomnia calls type of erros,validations

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: actions/templates.go [hunk 1]
```diff
@@ -110,6 +111,12 @@ func GetTemplates(c *gin.Context) {
 	})
 }
 
+type postTemplate struct {
+	Name    string `json:"name" form:"name" binding:"required,max=191"`
```
Comment (FilipNikolovski) [inline actions/templates.go:115]: @Dzalevski the json tag is redundant i think, we should only support `form`.

**Review by FilipNikolovski — CHANGES_REQUESTED** [review]

**Comment by FilipNikolovski** [comment]
I'm thinking we could upgrade the validator from v8 -> v9 by following this guide: https://medium.com/@seb.nyberg/better-validation-errors-in-go-gin-88f983564a3d so we can better utilize the FieldError (which in v9 is an interface) so we can customize the error message, like in the guide.

From the [upgrade guide](https://github.com/go-playground/validator/blob/v9/_examples/gin-upgrading-overriding/v8_to_v9.go) we'll need to set the binding.StructValidator to our new defaultValidator:

```go
package main

import (
	"reflect"
	"sync"

	"github.com/gin-gonic/gin/binding"
	"gopkg.in/go-playground/validator.v9"
)

type defaultValidator struct {
	once     sync.Once
	validate *validator.Validate
}

var _ binding.StructValidator = &defaultValidator{}

func (v *defaultValidator) ValidateStruct(obj interface{}) error {

	if kindOfData(obj) == reflect.Struct {

		v.lazyinit()

		if err := v.validate.Struct(obj); err != nil {
			return error(err)
		}
	}

	return nil
}

func (v *defaultValidator) Engine() interface{} {
	v.lazyinit()
	return v.validate
}

func (v *defaultValidator) lazyinit() {
	v.once.Do(func() {
		v.validate = validator.New()
		v.validate.SetTagName("binding")
                // Print JSON name on validator.FieldError.Field()
                v.validate.RegisterTagNameFunc(func(fld reflect.StructField) string {
                    name := strings.SplitN(fld.Tag.Get("form"), ",", 2)[0]
                    return name
                })
	})
}

func kindOfData(data interface{}) reflect.Kind {

	value := reflect.ValueOf(data)
	valueType := value.Kind()

	if valueType == reflect.Ptr {
		valueType = value.Elem().Kind()
	}
	return valueType
}
```

And we'll bind this new validator in our `mailbadger.go` file:

```go
package main
import (
  "github.com/gin-gonic/gin/binding"
...
)

func main() {
    binding.Validator = new(defaultValidator)
...
}
```

And in our new package `validation` we could have a fieldError struct that wraps the FieldError and implements the Stringer interface:

```go
package validation

import (
	"fmt"
	"strings"

	"gopkg.in/go-playground/validator.v9"
)

type FieldError struct {
	err validator.FieldError
}

func (q FieldError) String() string {
	var sb strings.Builder

	sb.WriteString("validation failed on field '" + q.err.Field() + "'")
	sb.WriteString(", condition: " + q.err.ActualTag())

	// Print condition parameters, e.g. oneof=red blue -> { red blue }
	if q.err.Param() != "" {
		sb.WriteString(" { " + q.err.Param() + " }")
	}

	if q.err.Value() != nil && q.err.Value() != "" {
		sb.WriteString(fmt.Sprintf(", actual: %v", q.err.Value()))
	}

	return sb.String()
}
```

Finally in our actions, when returning the error message we'll wrap the field error and return it:

```go
c.JSON(http.StatusBadRequest, validation.FieldError{fieldErr})
```

**Comment by Dzalevski** [comment]
@FilipNikolovski Again this will produce the same message as the actual implementation. Here is one example:
we need to validate that  a password is not longer then 8 chars. The validation will fail if pw is longer then 8 chars but message will be something like this: Failed on password tag MAX. With this message user doesn't know that he have pw that is longer that 8 chars :D . I think that we will need to check which validation failed and put our custom error I cannot see other implementation.

**Comment by FilipNikolovski** [comment]
@Dzalevski our String() implementation can be different, that's just an example. We can have a switch case for each `ActualTag` and set a custom message.

**Comment by Dzalevski** [comment]
@FilipNikolovski  yes that's what I said. Okay

File: validator/validator.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+type DefaultValidator struct {
+	once     sync.Once
+	validate *validator.Validate
+}
+
+var _ binding.StructValidator = &DefaultValidator{}
+
+func (v *DefaultValidator) ValidateStruct(obj interface{}) error {
+
```
Comment (FilipNikolovski) [inline validator/validator.go:19]: @Dzalevski remove empty lines.

**Review by FilipNikolovski — CHANGES_REQUESTED** [review]

**Review by FilipNikolovski — APPROVED** [review]

**Pull request merged**